File: suites/osprey-compute-kernels/reference/cases/ackermann/ackermann.c
/* Ackermann–Péter function — deep, non-tail mutual self-recursion. */
#include <stdint.h>
#include <stdio.h>

static int64_t add(int64_t a, int64_t b) { return a + b; }
static int64_t sub(int64_t a, int64_t b) { return a - b; }

static int64_t ack(int64_t m, int64_t n) {
    if (m == 0) {
        return add(n, 1);
    } else if (n == 0) {
        return ack(sub(m, 1), 1);
    } else {
        return ack(sub(m, 1), ack(m, sub(n, 1)));
    }
}

int main(void) {
    printf("%lld\n", (long long)ack(3, 10));
    return 0;
}

File: suites/osprey-compute-kernels/reference/cases/ackermann/ackermann.hs
-- Ackermann–Péter function — deep, non-tail mutual self-recursion.
add' :: Int -> Int -> Int
add' a b = a + b

sub' :: Int -> Int -> Int
sub' a b = a - b

ack :: Int -> Int -> Int
ack m n
  | m == 0 = add' n 1
  | n == 0 = ack (sub' m 1) 1
  | otherwise = ack (sub' m 1) (ack m (sub' n 1))

main :: IO ()
main = print (ack 3 10)
